### PR TITLE
Use `impl IntoIterator` to avoid clones

### PR DESCRIPTION
**Pull request by charliermarsh**
Reviving Jacob's PR to see if we can improve our error messages for conflicting dependencies within a single package.

### CODE REVIEW

**Comment by charliermarsh** [comment]
\cc @zanieb for visibility.

File: src/lib.rs [hunk 1]
```diff
@@ -103,8 +103,10 @@
 //!         &self,
 //!         package: &String,
 //!         version: &SemanticVersion,
-//!     ) -> Result<Dependencies<String, SemVS>, Infallible> {
-//!         unimplemented!()
+//!     ) -> Result<Dependencies<impl IntoIterator<Item = (String, SemVS)> + Clone>, Infallible>
+//!     {
+//!         unimplemented!();
+//!         Ok(Dependencies::Available([]))
```
Comment (zanieb) [inline src/lib.rs]: ?

File: src/solver.rs [hunk 2]
```diff
@@ -167,10 +167,10 @@ pub fn resolve<P: Package, VS: VersionSet, DP: DependencyProvider<P, VS>>(
...
                     continue;
                 }
-                Dependencies::Available(x) if x.contains_key(p) => {
+                Dependencies::Available(x) if x.clone().into_iter().any(|(d, _)| &d == p) => {
                     return Err(PubGrubError::SelfDependency {
                         package: p.clone(),
-                        version: v,
+                        version: v.clone(),
                     });
                 }
```
Comment (zanieb) [inline src/solver.rs:175]: Are any of these new clones problematic?

**Review by zanieb — APPROVED** [review]

File: src/solver.rs (hunk 2 above)
Comment (charliermarsh) [reply src/solver.rs:175]: Not for us, because we already use a vector here IIUC.

**Pull request merged**